### ISSUE TITLE
Help: KafkaConsumer auto commit doesn't appear to work, messages loop in queue forever

### ISSUE THREAD

**Issue by rjurney** [dpkp/kafka-python#1053]
I activate my KafkaConsumer like so:

```
from kafka import KafkaConsumer
kafka_consumer = KafkaConsumer(
  master_config["kafka"]["topic"],
  bootstrap_servers=master_config["kafka"]["brokers"],
  api_version=(0, 10),
  group_id="main",
  enable_auto_commit=True,
  security_protocol='SSL',
  ssl_check_hostname=True,
  ssl_cafile=master_config["kafka"]["ca.pem"],
  ssl_certfile=master_config["kafka"]["service.cert"],
  ssl_keyfile=master_config["kafka"]["service.key"]
)
```

Then I loop like so:

```
# Run the loop
for message_object in kafka_consumer:
  genetics.log_message(str(psutil.virtual_memory()))
  reload(genetics) # update to the latest code automatically
  genetics.process_message(message_object, ps_conn, s3_config)
```

However, messages keep replaying in my queue over and over. What am I to do? I've no idea how to address this. The queue has three partitions, which I'm using the consumer to read from three concurrent instances of my worker.

**Comment by dpkp** [comment]
I recommend enabling logging, perhaps even DEBUG level, and take a look at what is happening under the covers. "Replaying" messages might be caused by failures in group management which force resetting partition offsets.

**Comment by rjurney** [comment]
Do you mean in Kafka or in the consumer? If consumer, not sure how to do that. Can't find anything on logging in the API reference.

**Comment by dpkp** [comment]
python logging. So for example, you might do:
```
import logging
logging.basicConfig(level=logging.DEBUG)
```

**Comment by rjurney** [comment]
```
END OF LOOP (this is where my app ends its loop.

DEBUG:kafka.consumer.group:internal iterator timeout - breaking for poll
DEBUG:kafka.coordinator.consumer:Sending offset-commit request with {TopicPartition(topic='t_sne_request_redux', partition=1): OffsetAndMetadata(offset=9, metadata=''), TopicPartition(topic='t_sne_request_redux', partition=0): OffsetAndMetadata(offset=12, metadata='')} for group main to 3
DEBUG:kafka.conn:<BrokerConnection host=54.68.82.64/54.68.82.64 port=27503> Request 91: OffsetCommitRequest_v2(consumer_group='main', consumer_group_generation_id=211, consumer_id='kafka-python-1.3.2-2b3b1253-2266-4d04-b853-58eaeff82c08', retention_time=-1, topics=[(topic='t_sne_request_redux', partitions=[(partition=0, offset=12, metadata=''), (partition=1, offset=9, metadata='')])])
ERROR:kafka.coordinator:Heartbeat session expired - marking coordinator dead
```

Does this mean I need to update a setting to let the heartbeat last longer so the session doesn't expire?

**Comment by rjurney** [comment]
The work my process is doing takes ten minutes sometimes, but there is an error when I set the session timeout that high. I tried doing the work in a thread, but the main thread doesn't respond to heartbeat messages while it loops waiting on the worker thread to finish.

What is the correct way to do this for a long running work process? How can I transmit heartbeats back manually?

**Comment by rjurney** [comment]
Five minutes is the maximum timeout you can set :(

```
INFO:kafka.coordinator:(Re-)joining group main
DEBUG:kafka.coordinator:Sending JoinGroup (JoinGroupRequest_v0(group='main', session_timeout=600000, member_id='', protocol_type='consumer', group_protocols=[(protocol_name='range', protocol_metadata=b'\x00\x00\x00\x00\x00\x01\x00\x13t_sne_request_redux\x00\x00\x00\x00'), (protocol_name='roundrobin', protocol_metadata=b'\x00\x00\x00\x00\x00\x01\x00\x13t_sne_request_redux\x00\x00\x00\x00')])) to coordinator 3
DEBUG:kafka.conn:<BrokerConnection host=54.68.82.64/54.68.82.64 port=27503> Request 1: JoinGroupRequest_v0(group='main', session_timeout=600000, member_id='', protocol_type='consumer', group_protocols=[(protocol_name='range', protocol_metadata=b'\x00\x00\x00\x00\x00\x01\x00\x13t_sne_request_redux\x00\x00\x00\x00'), (protocol_name='roundrobin', protocol_metadata=b'\x00\x00\x00\x00\x00\x01\x00\x13t_sne_request_redux\x00\x00\x00\x00')])
DEBUG:kafka.conn:<BrokerConnection host=54.68.82.64/54.68.82.64 port=27503> Response 1: JoinGroupResponse_v0(error_code=26, generation_id=0, group_protocol='', leader_id='', member_id='', members=[])
ERROR:kafka.coordinator:Attempt to join group main failed due to fatal error: [Error 26] InvalidSessionTimeoutError: JoinGroupResponse_v0(error_code=26, generation_id=0, group_protocol='', leader_id='', member_id='', members=[])
Traceback (most recent call last):
  File "./worker.py", line 97, in <module>
    for message_object in kafka_consumer:
  File "/Users/rjurney/anaconda/lib/python3.5/site-packages/kafka/consumer/group.py", line 925, in __next__
    return next(self._iterator)
  File "/Users/rjurney/anaconda/lib/python3.5/site-packages/kafka/consumer/group.py", line 851, in _message_generator
    self._coordinator.ensure_active_group()
  File "/Users/rjurney/anaconda/lib/python3.5/site-packages/kafka/coordinator/base.py", line 273, in ensure_active_group
    raise exception # pylint: disable-msg=raising-bad-type
kafka.errors.InvalidSessionTimeoutError: [Error 26] InvalidSessionTimeoutError: JoinGroupResponse_v0(error_code=26, generation_id=0, group_protocol='', leader_id='', member_id='', members=[])
```

**Comment by rjurney** [comment]
Do you have any tips for me? The maximum timeout it will let me set is 5 minutes for the session. I don't understand why that is the case, and the code is very obscure so I'm not sure how to fork and override it.

**Comment by dpkp** [comment]
The maximum session timeout is set by the broker. Check your broker configuration if you think you need longer session timeout.

But honestly if you need 10 mins to process a single message, kafka may not be the right tool for you. Kafka is intended for high throughput streaming systems and is optimized accordingly. You aren't going to find it easy to work with if what you are looking for is a managed queue. rabbitMQ or redis may provide better semantics in that case.

**Comment by rjurney** [comment]
Thanks, I discovered this just before reading this and asked my provider to up the setting on the broker.

While Kafka is intended for high throughput, I don't think it is a bad fit for us. We process messages of a variety of lengths. Only the largest ones exceed the default timeout. 

My question is... why can't the consumer respond to heartbeats if I use a thread to conduct the work? This would enable the heartbeat to work... and how long it takes to process a message wouldn't matter. When I spawn a thread for the work, the heartbeats do not occur. This confuses me.

**Comment by rjurney** [comment]
Along these lines, I see these links regarding heartbeats in threads:

https://cwiki.apache.org/confluence/display/KAFKA/KIP-62%3A+Allow+consumer+to+send+heartbeats+from+a+background+thread
https://issues.apache.org/jira/browse/KAFKA-3888?src=confmacro

This was released in 0.10.1.0.

**Comment by jeffwidman** [comment]
There's an open ticket to add background heartbeats (KIP-62 functionality) to `kafka-python`: https://github.com/dpkp/kafka-python/issues/948

If you want to take a stab at it many people would appreciate it. In general `kafka-python` code is modeled after the official Java consumer so I'd start by looking at how they implemented it and try to copy the changes over.
